contracts/products/SoteriaCoverageProduct.sol

```diff
@@ -0,0 +1,681 @@
+     * @dev this is set to 0 to reset (default value is 0 in Solidity anyway)
+     */
+    mapping(address => uint256) internal _cooldownStart;
+
+    /// @notice The policyholder => Soteria account balance.
+    mapping(address => uint256) internal _accountBalanceOf; // Considered _soteriaAccountBalance name
+
+    /// @notice The policyholder => policyID.
+    mapping(address => uint256) internal _policyOf;
+
+    /// @notice The cover limit for each policy(policyID => coverLimit).
+    mapping (uint256 => uint256) internal _coverLimitOf;
+
+    /// @notice This mapping is created for the purpose of avoiding the unintended side effect where `user deactivates policy -> cover limit set to 0 -> minRequiredAccountBalance also set to 0 unintentionally because minRequiredAccountBalance = coverLimit * chargeCycle * maxRate -> user can withdraw all funds immediately after deactivating account
+    /// @dev This mapping is intended to mirror the _coverLimitOf mapping, except time between cooldown start and after cooldown completed
+    mapping (uint256 => uint256) internal _preDeactivateCoverLimitOf;
+
+    /// @notice The policy holder => reward points. Having a reward points mechanism enables `free` cover gifts and discounts for referrals.
+    mapping (address => uint256) internal _rewardPointsOf;
+
+    /// @notice PolicyID => Has referral code been used for this policyID?
+    mapping (uint256 => bool) internal _isReferralCodeUsed;
+
+    /***************************************
+    MODIFIERS
+    ***************************************/
+
+    modifier whileUnpaused() {
+        require(!_paused, "contract paused");
+        _;
+    }
+
+    /**
+     * @notice Constructs `Soteria` product.
+     * @param governance_ The governor.
+     * @param registry_ The [`Registry`](./Registry) contract address.
+     * @param domain_ The user readable name of the EIP712 signing domain.
+     * @param version_ The current major version of the signing domain.
+     */
+    constructor(
+        address governance_,
+        address registry_,
+        string memory domain_,
+        string memory version_
+    ) ERC721("Soteria Policy", "SOPT") Governable(governance_) EIP712(domain_, version_) {
+        require(registry_ != address(0x0), "zero address registry");
+        _registry = IRegistry(registry_);
+        require(_registry.get("riskManager") != address(0x0), "zero address riskmanager");
+    }
+    
+    /***************************************
+    FALLBACK FUNCTIONS
+    ***************************************/
+
+    /**
+     * @notice Fallback function to allow contract to receive *ETH*.
+    */
+    // solhint-disable-next-line 
+    receive() external payable nonReentrant {
+        _deposit(msg.sender, msg.value);
+    }
+
+    /**
+     * @notice Fallback function to allow contract to receive *ETH*.
+    */
+    fallback() external payable nonReentrant {
+        _deposit(msg.sender, msg.value);
```

Comment:
So, this smart contract manages everything in ETH as I see (accounts, deposits, withdrawals, etc). However we denominate covers in USD. Therefore, we could either keep things in ETH here and do accounting off-chain to charge respective dollar amounts in ETH, or need to switch this contract to be entirely operating USDC (or other stablecoin).

Reply:
For now let's keep things in ETH. The premium charger script  charges in ETH right now.

Reply:
Is it gonna be hard to switch to DAI?

Reply:
In the script, no. The premium is calculated in USD then it is converted to the ETH. But in smart contract side, there will be blood and tears:)